[GTK] Add support for explicit sync

carlosgcampos opened the pull request:
#### 1e569695d21e775da81a79111300975afe81b5f4
<pre>
[GTK] Add support for explicit sync
<a href="https://bugs.webkit.org/show_bug.cgi?id=277485">https://bugs.webkit.org/show_bug.cgi?id=277485</a>

Reviewed by Alejandro G. Castro.

In case of GTK we can&apos;t pass the fence to GTK to wait for it, but we can
just wait for it ourselves by polling the sync file. This way we avoid
the client wait in the web process on every rendered frame.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/glib/FenceMonitor.cpp: Added.
(WebKit::FenceMonitor::FenceMonitor):
(WebKit::FenceMonitor::~FenceMonitor):
(WebKit::FenceSource::~FenceSource):
(WebKit::FenceMonitor::ensureSource):
(WebKit::isFileDescriptorReadable):
(WebKit::FenceMonitor::addFileDescriptor):
* Source/WebKit/UIProcess/glib/FenceMonitor.h: Added.
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::useExplicitSync const):

Canonical link: <a href="https://commits.webkit.org/281744@main">https://commits.webkit.org/281744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93a77811fa43a7d5711b0c1069cd814cb6e8ff44

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60749 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64680 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11296 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49124 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7838 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62783 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52631 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29956 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34041 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10209 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55879 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66409 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4693 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9989 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56492 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4714 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56674 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3892 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9159 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35913 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36995 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38088 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36740 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas commented on Source/WebKit/UIProcess/glib/FenceMonitor.cpp:
```diff
@@ -0,0 +1,130 @@
+    nullptr, // prepare
+    nullptr, // check
+    // dispatch
+    [](GSource* base, GSourceFunc callback, gpointer userData) -> gboolean
+    {
+        auto& source = *reinterpret_cast<FenceSource*>(base);
+        g_source_remove_unix_fd(&source.base, source.tag);
+        source.tag = nullptr;
+        source.fd = { };
+
+        callback(userData);
+        return G_SOURCE_CONTINUE;
+    },
```
I think here it needs to check if the file descriptor is readable. Here's an example that does mostly the same: https://gitlab.gnome.org/GNOME/mutter/-/blob/main/src/wayland/meta-wayland-dma-buf.c?ref_type=heads#L868-926

carlosgcampos replied:
That was my inspiration :-) but here we only have one fd, so if the source is dispatched is because the fd became readable, no?

GeorgesStavracas replied:
I don't really know if the only event this FD can emit is when becoming readable...

carlosgcampos replied:
That's what I'm asking for in `g_source_add_unix_fd`, I guess I could check revents to be extra sure with `g_source_query_unix_fd` but I don't think we need to call `isFileDescriptorReadable` again

GeorgesStavracas replied:
Reading it again, I think this part of the patch is perfectly fine as is already. I don't think it needs change.

GeorgesStavracas reviewed (approved):
LGTM

benjamin-otte commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -143,6 +143,7 @@ std::unique_ptr<AcceleratedBackingStoreDMABuf> AcceleratedBackingStoreDMABuf::cr
 
 AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf(WebPageProxy& webPage)
     : AcceleratedBackingStore(webPage)
+    , m_fenceMonitor([this] { gtk_widget_queue_draw(m_webPage.viewWidget()); })
```
You need to wait with using the new dmabuf until the fence signals. Other parts of GTK may have queued a redraw in the meantime (going to the overview or resizing the window for example).

You also need to ensure nobody writes to the dmabuf until GTK has finished using it, which means not signaling the dmabuf until the GdkDmabufTexture's destroy notify runs.

carlosgcampos replied:
> You need to wait with using the new dmabuf until the fence signals. Other parts of GTK may have queued a redraw in the meantime (going to the overview or resizing the window for example).

That's what the patch does.

> You also need to ensure nobody writes to the dmabuf until GTK has finished using it, which means not signaling the dmabuf until the GdkDmabufTexture's destroy notify runs.

We don't exactly do this, we release the buffer right before rendering the next one, I'll check to switch to do it in the destroy notify callback.

alexgcastro reviewed (approved):
Great patch!

carlosgcampos committed:
[GTK] Add support for explicit sync
https://bugs.webkit.org/show_bug.cgi?id=277485

Reviewed by Alejandro G. Castro.

In case of GTK we can't pass the fence to GTK to wait for it, but we can
just wait for it ourselves by polling the sync file. This way we avoid
the client wait in the web process on every rendered frame.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/glib/FenceMonitor.cpp: Added.
(WebKit::FenceMonitor::FenceMonitor):
(WebKit::FenceMonitor::~FenceMonitor):
(WebKit::FenceSource::~FenceSource):
(WebKit::FenceMonitor::ensureSource):
(WebKit::isFileDescriptorReadable):
(WebKit::FenceMonitor::addFileDescriptor):
* Source/WebKit/UIProcess/glib/FenceMonitor.h: Added.
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::AcceleratedBackingStoreDMABuf):
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::useExplicitSync const):

Canonical link: https://commits.webkit.org/281744@main
